driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,322 @@
+import com.datastax.driver.core.exceptions.WriteTimeoutException;
+import com.google.common.collect.AbstractIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Chainable load balancing policy that will filter out hosts based on their errors rates.
+ * <p/>
+ * The policy bases its list of hosts on a child policy on which it will compute whether this host needs to included
+ * or excluded of the query plan.
+ * <p/>
+ * The policy uses a {@link LatencyTracker} that it will register automatically on the {@link Cluster}. This tracker
+ * will not take into account successful queries and only take note of the errors produced by queries. If the
+ * accounted number of errors for a host exceeds the threshold configured in the policy, the Host will be ignored
+ * for the query plan.
+ * <p/>
+ * The policy follows the builder pattern to be created, the {@link Builder} class can be created with
+ * {@link #builder} method.
+ */
+public class ErrorAwarePolicy implements ChainableLoadBalancingPolicy {
+
+    private static final Logger logger = LoggerFactory.getLogger(ErrorAwarePolicy.class);
+
+    private final LoadBalancingPolicy childPolicy;
+
+    private final int retryPeriodMillis;
+
+    PerHostErrorTracker errorTracker;
+
+    private ErrorAwarePolicy(Builder builder) {
+        this.childPolicy = builder.childPolicy;
+        this.retryPeriodMillis = builder.retryPeriod;
+        this.errorTracker = new PerHostErrorTracker(builder.errorsThreshold, builder.errorsFilter);
+    }
+
+    @Override
+    public LoadBalancingPolicy getChildPolicy() {
+        return childPolicy;
+    }
+
+    @Override
+    public void init(Cluster cluster, Collection<Host> hosts) {
+        childPolicy.init(cluster, hosts);
+        cluster.register(this.errorTracker);
+    }
+
+    @Override
+    public HostDistance distance(Host host) {
+        return childPolicy.distance(host);
+    }
+
+    @Override
+    public Iterator<Host> newQueryPlan(String loggedKeyspace, Statement statement) {
+        final Iterator<Host> childQueryPlan = childPolicy.newQueryPlan(loggedKeyspace, statement);
+
+        return new AbstractIterator<Host>() {
+
+            @Override
+            protected Host computeNext() {
+                long currentTime = System.currentTimeMillis();
+                while (childQueryPlan.hasNext()) {
+                    Host host = childQueryPlan.next();
+
+                    Long excludedTime = errorTracker.getExclusions().get(host);
+                    if (excludedTime == null) {
+                        // No entry in exclusions map means host hasn't crossed the threshold yet.
+                        return host;
+                    }
+
+                    if (currentTime - excludedTime > retryPeriodMillis) {
```

Comment:
One small problem here,  this just means there have been no errors in the retryPeriod, but the errorRate may still be greater than the inclusion threshold.  

I.E. consider that the retry period is 5 seconds (the default) and the inclusion threshold is 1 error/per sec.   If we get a flurry of errors for a host, say 200 errors in 5 seconds, the 1mrate will be > 1 for longer than 5 seconds, but in this case we would try the host right away.  Shouldn't we check the m1rate here?